src/AutoRest.CSharp.V3/Output/Builders/ClientBuilder.cs

```diff
@@ -307,67 +307,81 @@ private RestClientMethod BuildMethod(Operation operation, string clientName, IRe
                 // Ignore response body and headers for LROs as the ArmOperationHelpers figures out them dynamically
-                if (!operation.IsLongRunning)
-                {
-                    responseBody = BuildResponseBody(response);
-
-                    responseHeaderModel = BuildResponseHeaderModel(operation, response);
-                }
-
-                var responseCodes = new HashSet<int>(response.HttpResponse.IntStatusCodes);
-
                 // Long running operations can respond with both initial or final status code
-                if (operation.IsLongRunning)
-                {
-                    foreach (var statusCode in operation.LongRunningFinalResponse.HttpResponse.IntStatusCodes)
-                    {
-                        responseCodes.Add(statusCode);
-                    }
-                }
 
-                clientResponse = new Response(
-                    responseBody,
-                    responseCodes.ToArray(),
-                    responseHeaderModel
-                );
+                clientResponse.Add(new Response(
+                    null,
+                    operation.LongRunningInitialResponse.HttpResponse.IntStatusCodes.ToArray()
+                ));
+                clientResponse.Add(new Response(
+                    null,
+                    operation.LongRunningFinalResponse.HttpResponse.IntStatusCodes.ToArray()
+                ));
             }
             else
             {
-                // Special case for httpInfrastructure testServer swagger, service method always fails
-                clientResponse = new Response(null, Array.Empty<int>(), null);
+                foreach (var response in operation.Responses)
+                {
+                    clientResponse.Add(new Response(
+                        BuildResponseBody(response),
+                        response.HttpResponse.IntStatusCodes.ToArray()
+                    ));
+                }
+
+                responseHeaderModel = BuildResponseHeaderModel(operation);
             }
 
+            var responseType = ReduceResponses(clientResponse);
+
             return new RestClientMethod(
                 operationName,
                 BuilderHelpers.EscapeXmlDescription(operation.Language.Default.Description),
+                responseType,
                 request,
                 OrderParameters(methodParameters.Values),
-                clientResponse,
+                clientResponse.ToArray(),
+                responseHeaderModel,
                 new Diagnostic($"{clientName}.{operationName}", Array.Empty<DiagnosticAttribute>())
             );
         }
 
+        // Merges operations without response types types together
+        private CSharpType? ReduceResponses(List<Response> responses)
+        {
+            foreach (var typeGroup in responses.GroupBy(r=> r.ResponseBody))
```

Comment:
Does this not need to be a separate list if you're adding/removing items from it?

Reply:
`responses.GroupBy` consumes the entire list before starting to return results so another loop is not required.